### DISCUSSION TITLE
Why can I only detect only one locality ?

### DISCUSSION THREAD

**Discussion by phil-skillwon** [STEllAR-GROUP/hpx#6516] [Q&A] [answered]
My code is as follows:
```C++
hpx::this_thread::sleep_for(5s);
INFO("%s", __FUNCTION__);
std::vector<hpx::id_type> localities = hpx::find_all_localities();
if (localities.size() < 2) 
{
    ERROR("this program requires at least two localities");
    return;
}
```
When compiling HPX, I used HPX_WITH_PARCELPORT_TCP=ON. I ran the same program on two hosts in the same local network, but hpx::find_all_localities() cannot detect the locality on the other host. I checked the HPX documentation and added some command line parameters, such as:

HPX options (additionally allowed in an options file)
--hpx:worker
Run this instance in worker mode.

--hpx:console
Run this instance in console mode.

--hpx:connect
Run this instance in worker mode, but connecting late.

--hpx:run-agas-server
Run AGAS server as part of this runtime instance.

--hpx:run-hpx-main
Run the hpx_main function, regardless of locality mode.

--hpx:hpx arg
The IP address the HPX parcelport is listening on, expected format: address:port (default: 127.0.0.1:7910).

--hpx:agas arg
The IP address the AGAS root server is running on, expected format: address:port (default: 127.0.0.1:7910).

However, I still can't detect localities on other hosts. What should I do?

**Comment by hkaiser** [comment, marked as answer]
Please see here for an explanation of what's going on and how to resolve the problem: https://stackoverflow.com/questions/35367816/hpx-minimal-two-node-example-set-up

**Reply by phil-skillwon** [reply]
I really appreciate it.

**Reply by phil-skillwon** [reply]
> Please see here for an explanation of what's going on and how to resolve the problem: https://stackoverflow.com/questions/35367816/hpx-minimal-two-node-example-set-up

Now my HPX program can discover nodes on multiple hosts, but during testing I found that my functions cannot execute on multiple nodes. I referenced this example: https://github.com/STEllAR-GROUP/hpx/blob/master/examples/quickstart/hello_world_distributed.cpp, but I discovered that even when 2 nodes are detected, all tasks are executed on the root node. Tasks originally meant for node 1 are not executed on node 1, but instead run serially on node 0 after node 0's tasks finish. I'd like to ask, is this happening because I'm not using a job system like PBS or SLURM? I haven't used PBS or SLURM before.

**Reply by hkaiser** [reply]
I'd need to see your code to answer those questions, Can you share it?

**Reply by phil-skillwon** [reply]
> I'd need to see your code to answer those questions, Can you share it?

yes sure, here is my code:
main.cpp:
```C++
#include <iostream>
#include "Xlog.h"
#include "Iface.h"
// #include "hpx.hpp"
#include "hpx/hpx_start.hpp"
#include "hpx/version.hpp"
#include "hpx/init.hpp"
#include "TestHpx.h"

using namespace std;

static int hpx_main(int argc, char* argv[])
{
    TRACE("hpx_main function");

    testHpx();
    
    // Any HPX application logic goes here...
    return hpx::finalize();
}

int main(int argc, char **argv) 
{
    let ret = xlogInitFile("conf/LogConf.yaml");
    if (false == ret) 
    {
        cerr << "logger init failed." << endl;
        return -1;
    }

    INFO("hpx demostration running...");

    let hpxMajor = hpx::major_version();
    let hpxMinor = hpx::minor_version();
    let hpxPatch = hpx::subminor_version();
    INFO("hpx version: %d-%d-%d", hpxMajor, hpxMinor, hpxPatch);

    let hpxCfg = vector<string>();
    hpxCfg.push_back("hpx.handle_signals=0");
    hpxCfg.push_back("hpx.max_idle_loop_count=1000");
    hpxCfg.push_back("hpx.max_idle_backoff_time=1000");
    
    hpx::init_params initArgs;
    initArgs.cfg = std::move(hpxCfg);    

    ret = hpx::start(argc, argv, initArgs);
    if (false == ret) 
    {
        ERROR("hpx runtime init failed");
        return -3;
    }

    WARN("enter any key to exit");
    getchar();

    WARN("hpx demostration exiting...");

    return hpx::stop();
}
```

TestHpx.cpp:

```C++
#include "TestHpx.h"
#include "Xlog.h"
#include "DateTime.h"
#include "hpx/runtime.hpp"
#include <hpx/include/components.hpp>
#include <hpx/include/actions.hpp>
#include "hpx/barrier.hpp"
#include "hpx/include/lcos.hpp"
#include "hpx/include/async.hpp"
#include "hpx/runtime_distributed/get_num_localities.hpp"

using namespace std;

// compute-intensive task.
static int myTask(const int id) 
{
    TRACE("%s", __FUNCTION__);
    double result = 0.0, ts = 0.00;
    let ticker = Ticker();
    ticker.restart();
    const double computeTime = 20'000.00;
    while (true)
    {
        // perform a large number of floating-point operations to ensure the CPU is fully loaded.
        for (int i = 0; i < 1000000; ++i)
        {
            result += std::sin(i) * std::cos(i);
        }
        // check if it has been running for more than 'computeTime' milliseconds.
        ts = ticker.elapsed(Ticker::TimePrecision::PrecisionMs);
        if (ts > computeTime)
        {
            break;
        }
    }

    TRACE("task %d completed in %0.2f ms", id, ts);

    return id;
}

static int localityEntry(const uint32_t locId) 
{
    DEBUG("%s, locality id: %d", __FUNCTION__, locId);

    // get the number of worker OS-threads in use by this locality.
    const let numJobs = hpx::get_os_thread_count();
    std::vector<hpx::future<int>> futVec;
    for (std::size_t i = 0; i < numJobs; ++i) 
    {
        let myId = 100 * locId + i;

        let jobIndex = int16_t(i);
        let exec = hpx::execution::parallel_executor(hpx::threads::thread_priority::bound, hpx::threads::thread_stacksize::large,
            hpx::threads::thread_schedule_hint(hpx::threads::thread_schedule_hint_mode::thread, jobIndex));
        let fut = hpx::async(exec, myTask, i);

        futVec.push_back(std::move(fut));
    }

    hpx::wait_all(futVec);

    for (let& fut : futVec) 
    {
        DEBUG("result of task: %d", fut.get());
    }

    return locId;
}

HPX_PLAIN_ACTION(localityEntry, LocalityEntryAction);

void testHpx() 
{
    // hpx::this_thread::sleep_for(5s);
    INFO("%s", __FUNCTION__);

    hpx::error_code ec = hpx::make_success_code();
    std::vector<hpx::id_type> localities = hpx::find_all_localities(ec);
    if (hpx::error::success != ec.value()) 
    {
        ERROR("find_all_localities executed failed, %s", ec.get_message().c_str());
        return;
    }
    
    if (localities.size() < 2) 
    {
        ERROR("this program requires at least two localities");
        return;
    }

    for (const let& loc : localities) 
    {
        hpx::naming::gid_type gid = loc.get_gid();
        std::string address = hpx::get_locality_name(loc).get();
        std::uint32_t localityId = hpx::naming::get_locality_id_from_gid(gid);

        DEBUG("locality id: %d", localityId);
        DEBUG("locality address: %s", address.c_str());

        std::size_t os_thread_count = hpx::get_os_thread_count();
        DEBUG("number of OS threads: %d", os_thread_count);
    }

    std::vector<hpx::future<int>> futVec;
    for (const let& loc : localities) 
    {
        let numJobs = hpx::get_os_thread_count();
		
        hpx::naming::gid_type gid = loc.get_gid();
        std::string addr = hpx::get_locality_name(loc).get();
        std::uint32_t locId = hpx::naming::get_locality_id_from_gid(gid);
        DEBUG("locality gid %s loc id: %d, address: %s, num of pu: %d", gid.to_string().c_str(), locId, addr.c_str(), numJobs);

        let fut = hpx::async<LocalityEntryAction>(loc, locId);
        futVec.push_back(std::move(fut));
    }
    
    hpx::wait_all(futVec);

    for (let& fut : futVec) 
    {
        DEBUG("LocalityEntryAction result: %d", fut.get());
    }

}
```
Here is my startup command:
node0 (root node):
./HpxDemo.run --hpx:localities=2 --hpx:agas=192.168.6.115:7910 --hpx:hpx=192.168.6.115:7910 --hpx:threads 4
node1:
./HpxDemo.run --hpx:agas=192.168.6.115:7910 --hpx:hpx=192.168.6.117:7910 --hpx:worker --hpx:threads 4

**Reply by phil-skillwon** [reply]
I found the reason why myTask couldn't execute in parallel on 2 nodes. The blocking occurs on this call: 
std::string addr = hpx::get_locality_name(loc).get(); 
Could you please explain why this call causes the computation on locality 0 to execute first, and then the computation on locality 1 can continue?

**Reply by hkaiser** [reply]
> I found the reason why myTask couldn't execute in parallel on 2 nodes. The blocking occurs on this call: std::string addr = hpx::get_locality_name(loc).get(); Could you please explain why this call causes the computation on locality 0 to execute first, and then the computation on locality 1 can continue?

I don't think that this is the case. I can't think of any reason why the blocking call should prevent the remote invocation. Also, I assume the code above has a `#define let auto` hidden somewhere in the headers, hasn't it?

**Reply by phil-skillwon** [reply]
1. Indeed, #define let auto, so let is just auto.
2. I just conducted another test. When I commented out std::string addr = hpx::get_locality_name(loc).get();, both nodes started executing myTask almost simultaneously. If this line of code is not commented out, the root node (node 0) executes myTask first. After 20 seconds, when node 0 finishes execution, node 1 then starts executing myTask. I will show you my own log information from these two nodes shortly.

**Reply by phil-skillwon** [reply]
> > I found the reason why myTask couldn't execute in parallel on 2 nodes. The blocking occurs on this call: std::string addr = hpx::get_locality_name(loc).get(); Could you please explain why this call causes the computation on locality 0 to execute first, and then the computation on locality 1 can continue?
> 
> I don't think that this is the case. I can't think of any reason why the blocking call should prevent the remote invocation. Also, I assume the code above has a `#define let auto` hidden somewhere in the headers, hasn't it?

here is my log:
./HpxDemo.elf --hpx:localities 2 --hpx:agas 127.0.0.1:4001 --hpx:hpx 127.0.0.1:4001 --hpx:threads 2
Node0.log:
2024-07-02 19:51:33.591 - INFO - hpx demostration running...
2024-07-02 19:51:33.592 - INFO - hpx version: 1-10-0
2024-07-02 19:51:33.592 - INFO - hpx version string: 1.10.0
2024-07-02 19:51:52.406 - TRACE - hpx_main function
2024-07-02 19:51:52.406 - WARN - enter any key to exit
2024-07-02 19:51:52.406 - INFO - testHpx
2024-07-02 19:51:52.406 - DEBUG - locality id: 0
2024-07-02 19:51:52.406 - DEBUG - locality address: LinugearsXeon#0
2024-07-02 19:51:52.406 - DEBUG - number of OS threads: 2
2024-07-02 19:51:52.406 - DEBUG - locality id: 1
2024-07-02 19:51:52.406 - DEBUG - locality address: LinugearsXeon#1
2024-07-02 19:51:52.406 - DEBUG - number of OS threads: 2
2024-07-02 19:51:52.406 - DEBUG - locality gid 00000001000000000000000000000000 loc id: 0, num of pu: 2
2024-07-02 19:51:52.406 - DEBUG - localityEntry, locality id: 0
2024-07-02 19:51:52.407 - TRACE - myTask
2024-07-02 19:51:52.407 - TRACE - myTask
2024-07-02 19:52:12.407 - TRACE - task 0 completed in 20000.00 ms
2024-07-02 19:52:12.407 - TRACE - task 1 completed in 20000.00 ms
2024-07-02 19:52:12.407 - DEBUG - result of task: 0
2024-07-02 19:52:12.407 - DEBUG - result of task: 1
2024-07-02 19:52:12.407 - DEBUG - locality gid 00000002000000000000000000000000 loc id: 1, num of pu: 2
2024-07-02 19:52:32.409 - DEBUG - LocalityEntryAction result: 0
2024-07-02 19:52:32.409 - DEBUG - LocalityEntryAction result: 1

./HpxDemo.elf --hpx:worker --hpx:agas 127.0.0.1:4001 --hpx:hpx 127.0.0.1:4002 --hpx:threads 2
node1.log:
2024-07-02 19:51:52.382 - INFO - hpx demostration running...
2024-07-02 19:51:52.382 - INFO - hpx version: 1-10-0
2024-07-02 19:51:52.382 - INFO - hpx version string: 1.10.0
2024-07-02 19:51:52.406 - WARN - enter any key to exit
2024-07-02 19:52:12.408 - DEBUG - localityEntry, locality id: 1
2024-07-02 19:52:12.408 - TRACE - myTask
2024-07-02 19:52:12.408 - TRACE - myTask
2024-07-02 19:52:32.408 - TRACE - task 100 completed in 20000.00 ms
2024-07-02 19:52:32.408 - TRACE - task 101 completed in 20000.00 ms
2024-07-02 19:52:32.408 - DEBUG - result of task: 100
2024-07-02 19:52:32.408 - DEBUG - result of task: 101

Please note the timestamp when the myTask execution completes in node0.log and the timestamp when the myTask execution begins in node1.log.
node0.log:
024-07-02 19:51:52.406 - DEBUG - localityEntry, locality id: 0
2024-07-02 19:51:52.407 - TRACE - myTask
2024-07-02 19:51:52.407 - TRACE - myTask
2024-07-02 19:52:12.407 - TRACE - task 0 completed in 20000.00 ms
2024-07-02 19:52:12.407 - TRACE - task 1 completed in 20000.00 ms
2024-07-02 19:52:12.407 - DEBUG - result of task: 0
2024-07-02 19:52:12.407 - DEBUG - result of task: 1

node1.log:
2024-07-02 19:52:12.408 - DEBUG - localityEntry, locality id: 1
2024-07-02 19:52:12.408 - TRACE - myTask
2024-07-02 19:52:12.408 - TRACE - myTask
2024-07-02 19:52:32.408 - TRACE - task 100 completed in 20000.00 ms
2024-07-02 19:52:32.408 - TRACE - task 101 completed in 20000.00 ms
2024-07-02 19:52:32.408 - DEBUG - result of task: 100
2024-07-02 19:52:32.408 - DEBUG - result of task: 101

What is the internal implementation principle of hpx::get_locality_name(loc).get();? If its asynchronous execution is normal, is it possible that I missed some compilation options when compiling the HPX library?

**Reply by phil-skillwon** [reply]
I think this phenomenon might be caused by an error in my code above. I slightly modified the code from examples/quickstart/hello_world_distributed.cpp and pasted it into the hpx_main function, overwriting my original code. When running in the same environment, the asynchronous function call auto locName = hpx::get_locality_name(node).get(); no longer blocks.
here is my code, based on "hello_world_distributed.cpp":
```C++
///////////////////////////////////////////////////////////////////////////////
//  Copyright (c) 2007-2014 Hartmut Kaiser
//  Copyright (c) 2011 Bryce Adelstein-Lelbach
//
//  SPDX-License-Identifier: BSL-1.0
//  Distributed under the Boost Software License, Version 1.0. (See accompanying
//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
///////////////////////////////////////////////////////////////////////////////

// Including 'hpx/hpx_main.hpp' instead of the usual 'hpx/hpx_init.hpp' enables
// to use the plain C-main below as the direct main HPX entry point.
#include <hpx/config.hpp>
// #include <hpx/hpx_main.hpp>
#include <hpx/include/actions.hpp>
#include <hpx/include/components.hpp>
#include <hpx/include/lcos.hpp>
#include <hpx/include/parallel_executors.hpp>
#include <hpx/include/runtime.hpp>
#include <hpx/include/util.hpp>
#include <hpx/iostream.hpp>
#include "hpx/hpx_start.hpp"
#include "hpx/version.hpp"
#include "hpx/init.hpp"
#include <cstddef>
#include <cstdint>
#include <list>
#include <mutex>
#include <set>
#include <vector>
#include "hpx/thread.hpp"
#include <chrono>

using namespace std;

std::size_t hello_world_worker(std::size_t desired)
{
    std::size_t current = hpx::get_worker_thread_num();
    if (current == desired)
    {
        char const* msg = "hello world from OS-thread {1} on locality {2}\n";

        // hpx::util::format_to(hpx::cout, msg, desired, hpx::get_locality_id())
        //     << std::flush;

        printf("%s\n", __FUNCTION__);
        auto info = hpx::util::format(msg, desired, hpx::get_locality_id());
        printf("%s", info.c_str());
        hpx::this_thread::sleep_for(10s);
        printf("%s exit\n", __FUNCTION__);

        return desired;
    }

    return std::size_t(-1);
}

void hello_world_foreman()
{
    std::size_t const os_threads = hpx::get_os_thread_count();

    std::set<std::size_t> attendance;
    for (std::size_t os_thread = 0; os_thread < os_threads; ++os_thread)
        attendance.insert(os_thread);

    while (!attendance.empty())
    {
        std::vector<hpx::future<std::size_t>> futures;
        futures.reserve(attendance.size());

        for (std::size_t worker : attendance)
        {
            hpx::execution::parallel_executor exec(
                hpx::threads::thread_priority::bound);

            hpx::threads::thread_schedule_hint hint(
                hpx::threads::thread_schedule_hint_mode::thread,
                static_cast<std::int16_t>(worker));

            futures.push_back(
                hpx::async(hpx::execution::experimental::with_hint(exec, hint),
                    hello_world_worker, worker));
        }

        hpx::spinlock mtx;
        hpx::wait_each(hpx::unwrapping([&](std::size_t t) {
            if (std::size_t(-1) != t)
            {
                std::lock_guard<hpx::spinlock> lk(mtx);
                attendance.erase(t);
            }
        }),
            futures);
    }
}

HPX_PLAIN_ACTION(hello_world_foreman, hello_world_foreman_action)

int hpx_main(int argc, char* argv[])
{
    std::vector<hpx::id_type> localities = hpx::find_all_localities();
    std::vector<hpx::future<void>> futures;
    futures.reserve(localities.size());

    for (hpx::id_type const& node : localities)
    {
        auto locName = hpx::get_locality_name(node).get();
        printf("locality name: %s\n", locName.c_str());
        typedef hello_world_foreman_action action_type;
        futures.push_back(hpx::async<action_type>(node));
    }

    hpx::wait_all(futures);

    return hpx::finalize();
}

int main(int argc, char **argv) 
{
    printf("hpx demostration running...\n");  

    auto ret = hpx::start(argc, argv);
    if (false == ret) 
    {
        printf("hpx runtime init failed\n");
        return -3;
    }

    printf("enter any key to exit\n");
    getchar();
    printf("hpx demostration exiting...\n");

    return hpx::stop();
}
```

**Reply by phil-skillwon** [reply]
Based on the previous tests, I seem to have discovered an interesting phenomenon. I modified the myTask function, while keeping the rest of the code unchanged. The rest code is the main.cpp and TestHpx.cpp that I pasted earlier (https://github.com/STEllAR-GROUP/hpx/discussions/6516#discussioncomment-9931533). Now, my myTask function has three scenarios:

```C++
// compute-intensive task.
static int myTask(const int id) 
static int myTask(int id) 
{
    TRACE("%s", __FUNCTION__);
    double result = 0.0, ts = 0.00;
    let ticker = Ticker();
    ticker.restart();
    const double computeTime = 20'000.00;

#if 0
    while (true)
    {
        // perform a large number of floating-point operations to ensure the CPU is fully loaded.
        for (int i = 0; i < 1000000; ++i)
        {
            result += std::sin(i) * std::cos(i);
        }
        // check if it has been running for more than 'computeTime' milliseconds.
        ts = ticker.elapsed(Ticker::TimePrecision::PrecisionMs);
        if (ts > computeTime)
        {
            break;
        }
    }
#elif 1
    std::this_thread::sleep_for(10s);
#else 
    hpx::this_thread::sleep_for(10s);
#endif

    TRACE("task %d completed in %0.2f ms", id, ts);

    return id;
}
```

In the myTask function, the original while loop was simulating a compute-intensive task that would fully load the CPU. Later, I conducted multiple tests using std::this_thread::sleep_for or hpx::this_thread::sleep_for. Based on the test results, I can confirm:

1. Both the while loop and std::this_thread::sleep_for cause the asynchronous call  "std::string address = hpx::get_locality_name(loc).get(); "  to experience a long-term block.

2. hpx::this_thread::sleep_for does not cause hpx::get_locality_name(loc).get() to block. This scenario might be similar to what I observed in the "examples/quickstart/hello_world_distributed.cpp" test. The "examples/quickstart/hello_world_distributed.cpp" does not demonstrate how compute-intensive tasks are distributed and computed across CPU cores on multiple nodes.

These test results have made me even more confused. Could you please help me understand this?

**Reply by phil-skillwon** [reply]
@hkaiser Could you please take a moment to look at this code for me?
